db/view/view_update_generator.cc

```diff
@@ -86,7 +87,7 @@ future<> view_update_generator::start() {
                             ::mutation_reader::forwarding::no);
 
                     inject_failure("view_update_generator_consume_staging_sstable");
-                    auto result = staging_sstable_reader.consume_in_thread(view_updating_consumer(s, std::move(permit), *t, sstables, _as, staging_sstable_reader_handle));
+                    auto result = staging_sstable_reader.consume_in_thread(view_updating_consumer(s, std::move(permit), *t, std::move(excluded_sstables), _as, staging_sstable_reader_handle));
```

Comment:
very important that you excluded all staging despite only a subset will be processed in a round.

Reply:
@denesb For the future: I think make_reader_excluding..() is very generic and fragile. Rather than enforcing the correctness of "make reader excluding staging" to the caller, perhaps table should provide an interface which is less generic but more explicit: table::make_reader_excluding_staging_sstables(). Then we don't have to worry about what to exclude and we know that every call will be correct.

Reply:
Yes, I was thinking about that too. Especially that this is the only user of that method.

Reply:
> > Currently in each iteration of the view update generator loop, all accumulated staging sstables of all tables are processed, one table at a time. If for any reason staging sstables manage to pile up, this results in a "cliff" behaviour. All staging sstables for a table are processed at once and are all moved to the main sstable group only after all of them are processed. If for any reason the processing fails in the middle, the whole thing has to be restarted, with no incremental progress possible. This patch changes the loop so it processes sstables in small batches of 4. One iteration processes a batch from all tables, one table at a time. This makes process incremental. Even if a lot of staging sstable files managed to accumulate, a small batch has a much better chance of finishing without problems, leading to improved resiliency as if there is a failure, only the last batch has to be retried. As staging sstables are assumed to be disjoint, processing them in small batches doesn't create any duplicate work for us: we don't expect any partition to be found in more than a single sstable.
> > Fixes: #11744
> 
> Is the assumption correct? After repair the sstables will overlap.
> 
> Also, if we reshape before rebuild, we'll only have one sstable. Do we reshape? If not, should we?

We register sstables created by streaming / repair directly into view building, before reshape in off-strategy. I recall that with repair-based ops, there will be 1 sstable per vnode range per replica, (where they don't overlap?!). If we reshape before registering, it could happen that certain strategies will cause many files to have token overlapping, like with TWCS, which segregates by time window, not by token. View building uses partitioned_sstable_set unconditionally for building a combined reader. It's okay if staging sstables overlap by small degrees, as the incremental reader will not have to keep many readers opened at once.